File: OneCodeTeam/Create a soft keyboard (CSSoftKeyboard)/[C#]-Create a soft keyboard (CSSoftKeyboard)/C#/CSSoftKeyboard/UserInteraction/NativeMethods.cs
/****************************** Module Header ******************************\
 * Module Name:  NativeMethods.cs
 * Project:      CSSoftKeyboard
 * Copyright (c) Microsoft Corporation.
 * 
 * This class contains the structures that are used in SendInput method. 
 * 
 * 
 * This source is subject to the Microsoft Public License.
 * See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
 * All other rights reserved.
 * 
 * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
 * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
 * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
\***************************************************************************/


using System;
using System.Runtime.InteropServices;

namespace CSSoftKeyboard.UserInteraction
{
    internal static class NativeMethods
    {

        // The constants used in the INPUT structure.
        public const int INPUT_MOUSE = 0;
        public const int INPUT_KEYBOARD = 1;
        public const int INPUT_HARDWARE = 2;

        // The constants used in the KEYBDINPUT structure.
        public const int KEYEVENTF_EXTENDEDKEY = 0x0001;
        public const int KEYEVENTF_KEYUP = 0x0002;

        /// <summary>
        /// Used by SendInput to store information for synthesizing input events such 
        /// as keystrokes, mouse movement, and mouse clicks.
        /// http://msdn.microsoft.com/en-us/library/ms646270(VS.85).aspx
        /// </summary>
        [StructLayout(LayoutKind.Sequential)]
        public struct INPUT
        {
            /// <summary>
            /// INPUT_MOUSE    0
            /// INPUT_KEYBOARD 1
            /// INPUT_HARDWARE 2
            /// </summary>
            public int type;
            public NativeMethods.INPUTUNION inputUnion;
        }

        /// <summary>
        /// An INPUTUNION structure only contains one field. 
        /// http://msdn.microsoft.com/en-us/library/ms646270(VS.85).aspx
        /// </summary>
        [StructLayout(LayoutKind.Explicit)]
        public struct INPUTUNION
        {         
            [FieldOffset(0)]
            public NativeMethods.HARDWAREINPUT hi;
            [FieldOffset(0)]
            public NativeMethods.KEYBDINPUT ki;
            [FieldOffset(0)]
            public NativeMethods.MOUSEINPUT mi;
        }

        /// <summary>
        /// The information about a simulated hardware event.
        /// http://msdn.microsoft.com/en-us/library/ms646269(VS.85).aspx
        /// </summary>
        [StructLayout(LayoutKind.Sequential)]
        public struct HARDWAREINPUT
        {
            public int uMsg;
            public short wParamL;
            public short wParamH;
        }

        /// <summary>
        /// The information about a simulated keyboard event.
        /// http://msdn.microsoft.com/en-us/library/ms646271(VS.85).aspx
        /// </summary>
        [StructLayout(LayoutKind.Sequential)]
        public struct KEYBDINPUT
        {
            public short wVk;
            public short wScan;

            // KEYEVENTF_EXTENDEDKEY 0x0001
            // KEYEVENTF_KEYUP 0x0002
            // KEYEVENTF_SCANCODE 0x0008
            // KEYEVENTF_UNICODE 0x0004
            public int dwFlags;

            public int time;
            public IntPtr dwExtraInfo;
        }

        /// <summary>
        /// The information about a simulated mouse event.
        /// http://msdn.microsoft.com/en-us/library/ms646273(VS.85).aspx
        /// </summary>
        [StructLayout(LayoutKind.Sequential)]
        public struct MOUSEINPUT
        {
            public int dx;
            public int dy;
            public int mouseData;
            public int dwFlags;
            public int time;
            public IntPtr dwExtraInfo;
        }
    }


}

File: OneCodeTeam/Create a soft keyboard (VBSoftKeyboard)/[VB]-Create a soft keyboard (VBSoftKeyboard)/VB/VBSoftKeyboard/UserInteraction/NativeMethods.vb
'*************************** Module Header ******************************'
' Project:	    VBSoftKeyboard
' Copyright (c) Microsoft Corporation.
' 
' This class contains the structures and the constants which are used in
' the SendInput method. 
' 
' 
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
' All other rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'*************************************************************************'


Imports System.Runtime.InteropServices

Namespace UserInteraction
    Friend NotInheritable Class NativeMethods

        ' The constants used in the INPUT structure.
        Public Const INPUT_MOUSE As Integer = 0
        Public Const INPUT_KEYBOARD As Integer = 1
        Public Const INPUT_HARDWARE As Integer = 2

        ' The constants used in the KEYBDINPUT structure.
        Public Const KEYEVENTF_EXTENDEDKEY As Integer = &H1
        Public Const KEYEVENTF_KEYUP As Integer = &H2

        ''' <summary>
        ''' Used by SendInput to store information for synthesizing input events such 
        ''' as keystrokes, mouse movement, and mouse clicks.
        ''' http://msdn.microsoft.com/en-us/library/ms646270(VS.85).aspx
        ''' </summary>
        <StructLayout(LayoutKind.Sequential)>
        Public Structure INPUT
            ''' <summary>
            ''' INPUT_MOUSE    0
            ''' INPUT_KEYBOARD 1
            ''' INPUT_HARDWARE 2
            ''' </summary>
            Public type As Integer
            Public inputUnion As NativeMethods.INPUTUNION
        End Structure


        ''' <summary>
        ''' An INPUTUNION structure only contains one field. 
        ''' http://msdn.microsoft.com/en-us/library/ms646270(VS.85).aspx
        ''' </summary>
        <StructLayout(LayoutKind.Explicit)>
        Public Structure INPUTUNION
            <FieldOffset(0)>
            Public hi As NativeMethods.HARDWAREINPUT
            <FieldOffset(0)>
            Public ki As NativeMethods.KEYBDINPUT
            <FieldOffset(0)>
            Public mi As NativeMethods.MOUSEINPUT
        End Structure

        ''' <summary>
        ''' The information about a simulated hardware event.
        ''' http://msdn.microsoft.com/en-us/library/ms646269(VS.85).aspx
        ''' </summary>
        <StructLayout(LayoutKind.Sequential)>
        Public Structure HARDWAREINPUT
            Public uMsg As Integer
            Public wParamL As Short
            Public wParamH As Short
        End Structure

        ''' <summary>
        ''' The information about a simulated keyboard event.
        ''' http://msdn.microsoft.com/en-us/library/ms646271(VS.85).aspx
        ''' </summary>
        <StructLayout(LayoutKind.Sequential)>
        Public Structure KEYBDINPUT
            Public wVk As Short
            Public wScan As Short
            Public dwFlags As Integer
            Public time As Integer
            Public dwExtraInfo As IntPtr
        End Structure

        ''' <summary>
        ''' The information about a simulated mouse event.
        ''' http://msdn.microsoft.com/en-us/library/ms646273(VS.85).aspx
        ''' </summary>
        <StructLayout(LayoutKind.Sequential)>
        Public Structure MOUSEINPUT
            Public dx As Integer
            Public dy As Integer
            Public mouseData As Integer
            Public dwFlags As Integer
            Public time As Integer
            Public dwExtraInfo As IntPtr
        End Structure
    End Class

End Namespace
